Implemented baseline LoRA peft with FSDP integration, tested on one node.

jacobthebanana opened the pull request.

jacobthebanana committed:
Implemented baseline LoRA peft for one Nvidia GPU.

jacobthebanana committed:
Added support for saving lora adapters.
Added support for non-fsdp models.

jacobthebanana committed:
save_utils: added support for non-FSDP optimizers.
trainer: replaced clip_grad_norm_ with nn.utils.clip_grad_norm_ for lora compatibility.

jacobthebanana committed:
example_lora: highlighted current lora (non-fsdp) limitations.

jacobthebanana committed:
Added instructions on LoRA on one GPU.

jacobthebanana committed:
Added example script for launching lora.

jacobthebanana committed:
Revised instructions on LoRA on one GPU.

adil-a commented:
We need to implement at least one of DDP/FSDP as part of this PR.

jacobthebanana committed:
Implemented LoRA FSDP.
Also see https://github.com/facebookresearch/llama-recipes/blob/674b37ee66f59a7845cbc3868948f4d7fa69c679/src/llama_recipes/utils/fsdp_utils.py#L9

jacobthebanana commented:
This PR now features LoRA + FSDP, as requested.

Special thanks to @adil-a  for the pointer to [llama-recipes](https://github.com/facebookresearch/llama-recipes/) for the reference FSDP wrapper config.

jacobthebanana committed:
Reverted automatic formatter changes in README.md

jacobthebanana committed:
Eliminated non-FSDP logic from save_utils.
Set model path to local copy of llama-2-7b in example config.

jacobthebanana committed:
Moved lora config out of example config.yaml.

jacobthebanana commented:
Tested on 2x T4 16G:
- model: llama-2-7b
- eval_bs: 8
- train_bs: 8
- max_seq_len: **128**
- lora dimension: 8

jwilles commented:
We need to test this at scale before we can merge it, we need to understand the limits of common models on the A40s, A100s and capture those insights in the documentation for the rest of the community.

jacobthebanana commented:
Hello John, for sure. Could you elaborate a bit on the limits that we'd like to benchmark? A few things we might test include: 
- on the given GPU(s), does the model run without going out of memory?
- if the model fits, what is the training throughput? 

Also, what scale are we aiming for? At the moment, this implementation supports only models up to 7B because of CPU memory constraints. Shall we make a separate PR for that so that we can avoid changing many files at once, or would you prefer to address that within this PR as well?

jacobthebanana committed:
Implemented LoRA benchmarking logic for worker.

jacobthebanana committed:
model_utils: Refactored get_lora_model to reduce interface width. (this method no longer wraps load_model_and_tokenizer)
test_modelling: revised base model fixture scope since torch FSDP wrap is in-place.
launch_benchmark: added confirmation before launching.

jacobthebanana committed:
test_modelling: moved text output to data/.

jacobthebanana committed:
added example yaml config for lora benchmarking.

jacobthebanana committed:
launch_benchmark: marked qos flag as optional.

jacobthebanana committed:
launch_benchmark: added option to limit number of jobs launched.

jacobthebanana committed:
launch_benchmark: implemented torch profiler integration.

adil-a committed:
Merged changes from low CPU memory usage feature (#6) into jjt/lora-benchmarking

* added changes to implement low cpu mem usage feature

* implemented new ruff linting changes and ran a fix across files

jwilles commented:
Those two are a good starting point. Those are two of the first questions that researchers will ask. The memory constraint should have been addressed in #6 .

jacobthebanana committed:
Revised launch_benchmark.py to use new profiling path.

jacobthebanana committed:
Enabled automatic creation of data/trace folder.

jacobthebanana committed:
Added instructions for profiling tools.

jacobthebanana committed:
Merge remote-tracking branch 'origin/master' into jjt/lora-baseline

jacobthebanana committed:
Cleaned up duplicate imports from merge.

jacobthebanana committed:
Cleaned up duplicate imports from merge.

jacobthebanana committed:
Cleaned up parse_benchmark.py

jacobthebanana committed:
Integrated LoRA logic into llama_example.py.

jacobthebanana committed:
Moved lora_configs into train_parameters in config yaml. Adjusted docs/config.md accordingly.

jacobthebanana committed:
Revised handling of nproc-per-node in benchmark script.

jacobthebanana committed:
Included parameter_count info in benchmark output.

jacobthebanana committed:
Implemented basic util for parsing benchmarking output.

jacobthebanana commented:
The code hangs when combining low_cpu with lora. Possibly because LoRA is currently applied before FSDP, but the parameters are not materialized until after FSDP.

Without using LoRA:
- Instantiate model
- Apply FSDP Wrapper and materialize weights

LoRA + FSDP + low-cpu-ram (hangs on >1 GPUs):
- Instantiate Model
- Apply LoRA Wrapper
- Apply FSDP Wrapper and materialize weights

More investigation is needed.


---

Benchmark results (training token/second)

|                 |   Tesla T4 x1 |   NVIDIA A40 x1 |   Tesla T4 x2 |
|:----------------|--------------:|----------------:|--------------:|
| opt-350m        |      1208.52  |         2443.83 |           nan |
| Mistral-7B-v0.1 |       nan     |         1318.48 |           nan |
| Llama-2-7b-hf   |       246.372 |         1536.06 |           nan |
| Llama-2-13b-hf  |       nan     |         1068.81 |           nan |

--- 

Workaround for the moment, just to get the benchmark + profiler results: turn off low_cpu_mem_usage in the benchmark

jacobthebanana committed:
model_utils: Enabled  low_cpu_mem_usage in auto model from_pretrained by default.

jacobthebanana commented:
|                            |   Tesla T4 x7 |   Tesla T4 x2 |   NVIDIA A40 x2 |   Tesla T4 x6 |   Tesla T4 x8 |   NVIDIA A40 x1 |   Tesla T4 x5 |   NVIDIA A40 x3 |   Tesla T4 x4 |   Tesla T4 x1 |   NVIDIA A40 x4 |   Tesla T4 x3 |
|:---------------------------|--------------:|--------------:|----------------:|--------------:|--------------:|----------------:|--------------:|----------------:|--------------:|--------------:|----------------:|--------------:|
| opt-350m                   |      1894.86  |      2122.3   |        3521.92  |      2594.02  |      8326.63  |         2090.13 |      3555.7   |        5476.06  |      4245.36  |      1183.97  |         6779.7  |     2567.86   |
| Llama-2-7b-hf              |       325.917 |       114.551 |         738.724 |       307.592 |       414.597 |          nan    |       252.837 |        1071.1   |       217.184 |       249.344 |         1245.85 |      150.339  |
| Llama-2-13b-hf             |       nan     |      1178.05  |         400.686 |       153.307 |       nan     |          nan    |       130.075 |         376.147 |       110.61  |       nan     |          729.39 |       74.9184 |
| Mixtral-8x7B-Instruct-v0.1 |       nan     |       nan     |         nan     |       nan     |       nan     |          nan    |       nan     |         108.792 |       nan     |       nan     |          179.2  |      nan      |
| Mistral-7B-v0.1            |       317.002 |       101.245 |         661.11  |       304.622 |       402.589 |         1256.09 |       250.179 |         994.949 |       196.786 |       nan     |          nan    |      153.999  |

jacobthebanana committed:
launch_lora_benchmark.sh: implemented automatic identification of num_gpus.
lora-benchmark: switched
parse_benchmark: implemented option to specify benchmark artifact folder to load.

jacobthebanana committed:
requirements.txt: included accelerate to support low_cpu_mem loading.

jacobthebanana committed:
benchmark.py: adjusted BenchmarkingDataset to avoid StopIteration exception.

jacobthebanana commented:
Current workaround requires patching [transformer code](https://github.com/huggingface/transformers/blob/56b64bf1a51e29046bb3f8ca15839ff4d6a92c74/src/transformers/modeling_utils.py#L602) to set `assign_to_params_buffers` to True. Need to find a workaround that does not require patching.

jacobthebanana committed:
benchmark.py: added env var flag to toggle export_trace

jacobthebanana committed:
parse_benchmark: included profiler table in output file.
launch_benchmark: automated folder creation.
launch_lora_benchmark: included model info in slurm output.

jacobthebanana committed:
get_lora_model_from_base_model: enabled peft for models loaded via low_cpu_mem.
More investigation might be needed.

jacobthebanana committed:
model_utils: revised dtype handling for peft-wrapped models.

jacobthebanana commented:
|                            |   NVIDIA A40 x1 |   NVIDIA A40 x2 |   NVIDIA A40 x4 |   Tesla T4 x1 |   Tesla T4 x2 |   Tesla T4 x4 |   Tesla T4 x8 |
|:---------------------------|----------------:|----------------:|----------------:|--------------:|--------------:|--------------:|--------------:|
| Llama-2-13b-hf             |         1025.99 |         400.097 |         714.779 |        nan    |       5984.12 |       2184.21 |       4872.94 |
| Llama-2-7b-hf              |         1224.02 |         732.424 |        1291.46  |       6546.46 |       1484.95 |       2725.24 |       5163.52 |
| Mixtral-8x7B-Instruct-v0.1 |          nan    |         nan     |         194.435 |        nan    |        nan    |        nan    |        nan    |
| opt-350m                   |         2080.08 |        3157.4   |        5762.44  |       1084.22 |       2253.42 |       4331.24 |       7348.2  |

I'm trying to make sure I'm not mistaken here: Is the fact that we can now LoRA fine-tune Mixtral-8x7B using 4x A40 on a node with no more than 176GB of CPU RAM ([source](https://support.vectorinstitute.ai/Vaughan_slurm_changes)) sufficient to suggest that we've achieved the goal of this PR?

jacobthebanana committed:
parse_benchmark: implemented sorting of profiler table output.
launch_benchmark: revised default run time limit.

jacobthebanana committed:
Merged example_lora into examples/llama_example.pu

jacobthebanana committed:
Added instructions related to parse_benchmark

jacobthebanana committed:
parse_benchmark: implemented aggregation across repeated metrics.

adil-a commented on configs/config.yaml:
```diff
@@ -20,6 +20,14 @@ train_parameters:
   use_flash_attention: True
   low_cpu_mem_usage: True
 
+  # # Uncomment below to enable LoRA
```
pedantic but comment needs only one leading hashtag

jacobthebanana replied:
Resolved. Thanks for the feedback.

adil-a commented on README.md:
```diff
@@ -63,9 +63,12 @@ We have provided an example script to show what a regular workflow would look li
+
+We provide an additional example of parameter-efficient fine-tuning (PEFT) using LoRA and FSDP. Use the [`examples/launch_lora.sh`](examples/launch_lora.sh) to launch your job on the cluster.
+
+At the end of the training, the LoRA adapter folder will be saved in your output directory. This folder can be loaded directly through the `peft` library through the 
```
We should also include some of the benchmarking results from LoRA FSDP into the training optimizations markdown file (maybe dump the whole table in there and add a few comments on it).

adil-a replied:
And also mention in these lines that you've put the table in that markdown file.

jacobthebanana replied:
Deleted this section in README.md since we are now reusing the main launch script for LoRA (PEFT is specified via the shared config.yaml). Feel free to reopen if you'd like to mention the LoRA support here.

adil-a commented on examples/launch.sh:
```diff
@@ -23,4 +23,4 @@ export LOGLEVEL=INFO
 # export CUDA_LAUNCH_BLOCKING=0
 
-torchrun --nnodes=1 --nproc-per-node=4 llama_example.py --yaml_path ../configs/config.yaml
+torchrun --nnodes=1 --nproc-per-node=${SLURM_STEP_GPUS} llama_example.py --yaml_path ../configs/config.yaml
```
Good change, but I think the correct environment variable to use is `SLURM_GPUS_ON_NODE` as seen here: https://slurm.schedmd.com/sbatch.html#OPT_SLURM_GPUS_ON_NODE

The line will then be `torchrun --nnodes=1 --nproc-per-node=$SLURM_GPUS_ON_NODE ...`

jacobthebanana replied:
Seems that Vector Vaughan didn't properly support this. Let's stick with `nvidia-smi -S | wc -l` for now. I will make a ticket with ops and CC everyone here.
```
jacobtian@gpu063:~/vectorlm-prod$ export |grep SLURM_STEP
declare -x SLURM_STEPID="0"
declare -x SLURM_STEP_GPUS="4"
declare -x SLURM_STEP_ID="0"
declare -x SLURM_STEP_LAUNCHER_PORT="42115"
declare -x SLURM_STEP_NODELIST="gpu063"
declare -x SLURM_STEP_NUM_NODES="1"
declare -x SLURM_STEP_NUM_TASKS="1"
declare -x SLURM_STEP_TASKS_PER_NODE="1"
jacobtian@gpu063:~/vectorlm-prod$ squeue --me
   JOBID     NAME     USER PARTITION NODES TRES_PER_N ST          START_TIME         TIME NODELIST(REASON)
12297822     bash jacobtia      t4v2     1 gres:gpu:1  R 2024-04-07T16:00:58     19:06:17 gpu063
```

jacobthebanana replied:
Resolved in 56cb750.

adil-a commented on examples/launch_lora.sh:
```diff
@@ -0,0 +1,26 @@
+export PYTHONFAULTHANDLER=1
+# export CUDA_LAUNCH_BLOCKING=0
+
+torchrun --nnodes=1 --nproc-per-node=2 example_lora.py --yaml_path configs/config-lora.yaml
```
Similar comment, use the `$SLURM_GPUS_ON_NODE` variable to set the number of processes for torchrun.

adil-a commented on examples/launch_lora_one_gpu.sh:
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
+#SBATCH --job-name=llama7b-2-lora
+#SBATCH --nodes=1
+#SBATCH --mem=0
```
For single GPU, we might not want to swallow the entire node's CPU memory. Maybe change this to something more reasonable for 7B e.g. `64G`?

jacobthebanana replied:
Resolved in 2d0dfa8. These lora-specific files are no longer required.

adil-a commented on examples/llama_example.py:
```diff
@@ -67,9 +74,14 @@ def main(config: Config) -> None:
         training_args.low_cpu_mem_usage,
     )
 
+    lora_peft_config = getattr(config.train_parameters, "lora_peft_config", None)
```
Use `config.train_parameters.get("lora_peft_config")`. More pythonic way of dealing with, what essentially is, a dictionary.

jacobthebanana replied:
Resolved in 729267a. Thanks for pointing out that `box.Box` also support these methods.

adil-a commented on examples/llama_example.py:
```diff
@@ -67,9 +74,14 @@ def main(config: Config) -> None:
+    if lora_peft_config is not None:
+        model = get_lora_model_from_base_model(model, lora_peft_config)
+
+    decoder_layer_module = get_submodule_by_pattern(model, r"DecoderLayer$")
```
Is the import of `LlamaDecoderLayer` still being used after this? If not, feel free to remove.

adil-a commented on profiling/README.md:
```diff
@@ -0,0 +1,20 @@
+
+# The launcher script will print a list of
+# SLURM commands it plans to run. Press ENTER
+# to accept and automatically invoke the comands.
```
comands -> commands

adil-a commented on profiling/benchmark.py:
Please make sure you run `ruff` on this file and all other files you've created. It does not seem to be formatted correctly (maybe just do a `ruff check profiling/` and go through the issues). Make sure that the formatting rules I've excluded in `pyproject.toml` are actually being excluded in the problems that ruff throws at you. Last time all of my `__init__.py` files got wiped because they weren't excluded.

jacobthebanana replied:
Resolved. Thanks for the feedback.

adil-a commented on profiling/benchmark.py:
```diff
@@ -0,0 +1,360 @@
+
+
+# unix timestamp
+launch_time = time.time()
```
Can we put these under the `main` function?

jacobthebanana replied:
Since `write_metrics` is used strictly within the context of benchmarking, the extra complexity involved in doing so doesn't seem to justify doing so. Please reopen if you think otherwise.

adil-a commented on profiling/benchmark.py:
```diff
@@ -0,0 +1,360 @@
+os.makedirs("data/benchmark", exist_ok=True)
+os.makedirs("data/trace", exist_ok=True)
+output_path = "data/benchmark/{}.jsonl".format(launch_time)
+profiler_output_path = "data/trace/{}.json".format(launch_time)
```
Is there a reason we don't just make them both be either jsonl or json files? Also, let's stick to f-strings i.e. `output_path = f"data/benchmark/{launch_time}.jsonl"`.

jacobthebanana replied:
Chrome profiler viewer trace requires a json of its own. So need to use json for this one.
Rationale behind using jsonlines for throughput records: jsonlines can be flushed multiple times without going to the end of the file, making the output readable via standard tools even if the profiler process crashed.

Please open a new ticket/issue if you'd like to update the behavior of the chrome viewer integration.

String formatting convention is resolved in 37f5dec.

adil-a commented on profiling/benchmark.py:
```diff
@@ -0,0 +1,360 @@
+
+
+@contextlib.contextmanager
+def track_time(task_name: str, extra_info: Dict[str, Any] = {}):
```
`ruff` will catch this but missing docstring. Please fix.

jacobthebanana replied:
Resolved in 37f5dec. Thanks for pointing this out.

adil-a commented on profiling/benchmark.py:
```diff
@@ -0,0 +1,360 @@
+    checkpointed_epoch = 0
+
+    # See pytorch.org/tutorials/recipes/recipes/profiler_recipe.html
+    with torch.profiler.profile(
```
Should we allow the ability to save the whole tensorboard viewable trace? I know you output the chrome trace, but the tensorboard trace contains the chrome trace as well (albeit the file size is much larger). You just need to pass in something like `on_trace_ready=torch.profiler.tensorboard_trace_handler('./log'),`.

jacobthebanana replied:
We are asked to keep this ticket focused on LoRA integration and basic throughput benchmarking. Let's open a new ticket/issue if you find this feature important.

adil-a commented on profiling/benchmark.py:
```diff
@@ -0,0 +1,360 @@
+    write_metrics("slurm_info", get_slurm_env())
+
+    profiler_schedule = torch.profiler.schedule(
+        skip_first=10, wait=5, warmup=1, active=3, repeat=2
```
Is it possible to let these be these configurable arguments instead? `warmup` (maybe 10) should also be larger because the torch profiler has some overhead, and `wait` (maybe 20) should be larger because training takes a while to come to a consistent iteration speed. Rest are fine.

adil-a commented on profiling/benchmark.py:
```diff
@@ -0,0 +1,360 @@
+            train_dl_iterator = iter(dataset.train_dataloader)
+            for _ in tqdm(
+                # range(len(dataset.train_dataloader)),
+                range(7 * 13),
```
Where are these numbers from? Avoid hardcoding

jacobthebanana replied:
Resolved in 37f5dec.

adil-a commented on profiling/parse_benchmark.py:
```diff
@@ -0,0 +1,119 @@
+
+import pandas
+
+V = TypeVar("V", Dict, str, int)
```
Can this be a more informative variable name? i.e. is it defining some sort of state?

jacobthebanana replied:
Resolved. Thanks for the feedback.

adil-a commented on requirements.txt:
```diff
@@ -1,3 +1,4 @@
+accelerate
```
Where is this dependency introduced?

jacobthebanana replied:
Resolved in c1b31c4b.

adil-a commented on vectorlm/trainer.py:
```diff
@@ -48,13 +58,16 @@ class Trainer:
         max_steps: An integer maximum number of training steps for this run.
         saving_steps: An integer for how often we save.
 
```
Include `timer_handle` in the class attributes as well.

adil-a commented on vectorlm/trainer.py:
```diff
@@ -117,6 +133,7 @@ def prepare_trainer(
             optimizer: The training optimizer.
             lr_scheduler: The LR scheduler.
 
+
```
Are these new lines required by ruff?

jacobthebanana replied:
Applied ruff formatting.
```
$ ruff check vectorlm profiling examples && echo $?
warning: `one-blank-line-before-class` (D203) and `no-blank-line-before-class` (D211) are incompatible. Ignoring `one-blank-line-before-class`.
warning: `multi-line-summary-first-line` (D212) and `multi-line-summary-second-line` (D213) are incompatible. Ignoring `multi-line-summary-second-line`.
0
```

adil-a commented on vectorlm/trainer.py:
```diff
@@ -242,26 +268,42 @@ def train_step(self, batch: dict[str, torch.Tensor], epoch: int) -> float:
+                    out = self.model(**batch)
+                    tr_step_loss = out.loss
+                    (tr_step_loss / self.gas).backward()
+                    torch.nn.utils.clip_grad_norm_(
```
crucial revert. FSDP uses its own `clip_grad_norm` function as seen here: https://pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.FullyShardedDataParallel.clip_grad_norm_

Please change back to `self.model.clip_grad_norm_(self.config.max_grad_norm)`

jacobthebanana replied:
Resolved in 5babf6b as FSDP is always enabled (not DDP.)

adil-a commented on vectorlm/trainer.py:
```diff
@@ -242,26 +268,42 @@ def train_step(self, batch: dict[str, torch.Tensor], epoch: int) -> float:
         ids = batch.pop("id").to(torch.cuda.current_device())
         batch["input_ids"] = batch["input_ids"].type(torch.LongTensor)
         batch["labels"] = batch["labels"].type(torch.LongTensor)
+        batch = {k: v.to(torch.cuda.current_device()) for k, v in batch.items()}
```
Is this necessary? I think when tensors are created in the data collator they are on the correct device.

jacobthebanana replied:
Indeed- FSDP handles tensor placement, so this is required only for pure DDP. Fixed.

adil-a commented on vectorlm/trainer.py:
```diff
@@ -242,26 +268,42 @@ def train_step(self, batch: dict[str, torch.Tensor], epoch: int) -> float:
             tr_step_loss = out.loss
             (tr_step_loss / self.gas).backward()
-            self.model.clip_grad_norm_(self.config.max_grad_norm)
+            torch.nn.utils.clip_grad_norm_(
```
As asked above, revert.

jacobthebanana replied:
Resolved in 5babf6b.

adil-a commented on vectorlm/trainer.py:
```diff
@@ -291,9 +334,14 @@ def eval_step(self, epoch: int) -> float:
                 batch["labels"] = batch["labels"].type(torch.LongTensor)
-                out = self.model(**batch)
-                eval_loss += out.loss
+                batch = {k: v.to(torch.cuda.current_device()) for k, v in batch.items()}
```
As above, I believe this line is unnecessary.

jacobthebanana replied:
Resolved by deleting this line since we are exclusively running FSDP.

adil-a commented on vectorlm/utils/model_utils.py:
Seems ruff hasn't been run on this file either. IIRC `Dict` is no longer the preferred typing annotator. Instead use `dict[...]` in newer Python versions.

jacobthebanana replied:
Resolved in 37f5dec.

jacobthebanana replied:
```
$ ruff check vectorlm/utils/model_utils.py && echo $?
warning: `one-blank-line-before-class` (D203) and `no-blank-line-before-class` (D211) are incompatible. Ignoring `one-blank-line-before-class`.
warning: `multi-line-summary-first-line` (D212) and `multi-line-summary-second-line` (D213) are incompatible. Ignoring `multi-line-summary-second-line`.
0
```

adil-a commented on vectorlm/utils/model_utils.py:
```diff
@@ -25,6 +27,57 @@
+
+
+def get_lora_model_from_base_model(
+    base_model: nn.Module, peft_config_dict: Dict
```
dict annotation here also doesn't mention what are the types inside.

jacobthebanana replied:
Replaced with `dict[str, Any]` following the example of `fsdp_config`. Feel free to open a new refactor ticket if we want to make all config `dict` explicit (would indeed be quite helpful for programmers.)

adil-a commented on vectorlm/utils/model_utils.py:
```diff
@@ -142,7 +197,7 @@ def load_model_and_tokenizer(
 def fsdp_config(
     use_mp: bool,
-    layer_to_wrap: nn.Module,
+    model: nn.Module,
```
I am not a fan of this change because it now requires the use of `fsdp_auto_wrap_policy` which creates a weird dependency with accelerate code (and its subsequent environment variables). I am in strong favor of reverting this change to what it was because it strictly uses PyTorch. See [here](https://github.com/meta-llama/llama-recipes/blob/c8f4bdac4107856e204b281e4fb611dc4e2de5dc/src/llama_recipes/utils/fsdp_utils.py#L6) for a better implementation of what you might be trying to get at. We should stick to using Torch FSDP wrappers.

The two functions are very similar, but the one you're using starts involving accelerate which we're trying to avoid.

jacobthebanana replied:
Resolved in c1b31c4. Also eliminated implicit dependency on `accelerate` through `peft.utils.other.fsdp_auto_wrap_policy`. Thanks for the suggestions.

adil-a commented on vectorlm/utils/model_utils.py:
```diff
@@ -226,9 +277,14 @@ def shard_model(
     fsdp_cfg = fsdp_config(
-        use_mp, layer_to_wrap, strategy, local_rank, low_cpu_mem_usage,
+        use_mp,
+        model,
```
Following previous comment, needs to be changed as well.

adil-a commented on vectorlm/utils/save_utils.py:
```diff
@@ -275,4 +275,4 @@ def load_scheduler(
     state_dict = torch.load(input_scheduler_file)
     scheduler.load_state_dict(state_dict)
-    print(f"Scheduler state loaded from {input_scheduler_file}")
+    print(f"Scheduler state loaded from {input_scheduler_file}")
```
I believe new lines at the end of files are required by ruff

jacobthebanana replied:
Applied ruff formatting as requested.

adil-a commented on vectorlm/trainer.py:
Can you comment on what timer_placeholder does as default behavior? Seems like an obvious question but we shouldn't be seeing any slowdowns by unnecessarily calling it correct (since the default behavior is to call it at each training step for example)?

jacobthebanana replied:
`timer_placeholder` has been eliminated in afb321c.

adil-a reviewed (changes requested):
Changes requested.

Aside from explicit comments, please run `ruff` across all the files you have changed. Additionally, there are researchers using our code so we need to ensure that we're upholding the same training throughput before merging into main. In other words, we need to make sure we're stil getting ~27k tokens/s across 8 A100s using ctx length of 4096, bs=8, hybrid sharding. Some changes have been made to the main trainer and model util files so I'd like to make sure we're holding the same standard before finalizing.

Great job on the PR otherwise :)

jacobthebanana committed:
Implemented non-LoRA profiling and benchmarking.

jacobthebanana committed:
Various static typechecking and formatting fixes.

jacobthebanana committed:
Implemented restoring LoRA train state from filesystem.
During training the adapter weights are saved to and loaded from the filesystem. The base model weights are loaded separately.
Revised reference to optim_state_dict_to_load in load_optimizer.

jacobthebanana committed:
Included train step number in LoRA adapter output path.

jacobthebanana commented:
This branch now supports resuming LoRA training from preemption as many times as necessary. During training, the LoRA adapter weights are saved in place of the weights of the base model. The optimizer state dict is also relatively compact, as seen in the following example (gemma-2b):

```
total 5.5M
 647 Apr 15 19:02 adapter_config.json
1.8M Apr 15 19:02 adapter_model.safetensors
7.1K Apr 15 19:02 meta_data.pkl
3.7M Apr 15 19:02 optimizer_rank0.bin
5.0K Apr 15 19:02 README.md
1.0K Apr 15 19:02 scheduler_rank0.bin
```

![Resuming LoRA training from filesystem](https://github.com/VectorInstitute/vectorlm/assets/50071502/3d22e768-d305-4649-a4c9-835a0ae7af63)


More training metrics are available in [this wandb report](https://wandb.ai/jacobthebanana/vector-lm-verify/reports/2024-04-15-VectorLM-LoRA-Resume-from-Preemption--Vmlldzo3NTU4NjAx?accessToken=kowsnm6fumtywa0ueizm279nn90pd3swziaa243wxwj7endnlwdaaksfcaqlqjci).

jacobthebanana commented:
> Changes requested.
> 
> Aside from explicit comments, please run `ruff` across all the files you have changed. Additionally, there are researchers using our code so we need to ensure that we're upholding the same training throughput before merging into main. In other words, we need to make sure we're stil getting ~27k tokens/s across 8 A100s using ctx length of 4096, bs=8, hybrid sharding. Some changes have been made to the main trainer and model util files so I'd like to make sure we're holding the same standard before finalizing.

Thanks for the comments, Adil, and for reviewing the pull request. 

As requested, I'm trying to spin up experiments to find out if the changes I made negatively affected the throughput of the non-LoRA training. Before we reach out to ops for 8 A100 GPUs, would you mind elaborating on the steps you followed to obtain those throughput number? Additionally, given the inherent variance in training throughput, it will be great if you can share the min, 25%, median, 75%, and max readings of training throughput.

Please be sure to provide the following:
- Dataset
  - Either a version that is already processed and ready for llama_example, or
  - The original dataset, the preprocessing configuration yaml, and the script you ran to tokenize the dataset.
- Training configuration yaml
- SLURM script and the command you ran to launch the job.
- Output of `pip freeze`
- `modules` that you loaded when creating your virtual environment (if any) and when running the experiments, especially modules related to CUDA.

Additionally, as you know, the recent NVIDIA driver update on the cluster seemed to have substantially reduced the training throughput. Hence, we might not obtain exactly the same throughput as before the driver upgrade. Nevertheless, I'm sure simply comparing our throughput on this commit vs the main branch will provide enough info to draw conclusion about whether this PR affects non-LoRA throughput.

jacobthebanana committed:
Added reference throughput table to documentation.

jacobthebanana committed:
Added unit description to reference throughput table.
Applied markdown formatting via prettier.

jacobthebanana committed:
Added unit description to reference throughput table.
Applied markdown formatting via prettier.

jacobthebanana committed:
Benchmark: added option to override max_length of pre-trained model.

jacobthebanana committed:
Deleted unused `accelerate` dependency from requirements.txt

jacobthebanana committed:
Benchmark: added comment on max_length.

jacobthebanana committed:
Benchmark: added comment on batch size.

jacobthebanana committed:
Benchmark: added option to override batch size.

jacobthebanana committed:
Benchmark throughput documentation: revised word choices.

jacobthebanana committed:
Moved profiling-tracking logic out of Trainer.

jacobthebanana committed:
Eliminated hasattr check related to no_sync since FSDP is always enabled.

jacobthebanana committed:
Replaced peft fsdp_auto_wrap_policy to eliminate implicit `accelerate` dependency.
Eliminated redundant bfloat16 type conversion.
Fixed scope of placeholder for `is_peft_adapter_restored`.

jacobthebanana committed:
Configured LoRA auto-wrap policy as off by default- enable the policy only when LoRA is required.

jacobthebanana committed:
Revised punctuation in lora_requires_grad_policy_fn.

jacobthebanana committed:
Renamed declarative `enable_lora` with descriptive `is_lora_enabled`.

jacobthebanana commented:
The most recent pull request should have enabled full-rank fine-tuning on this branch to match the same level of throughput as the main branch.

Link to [output](https://gist.github.com/jacobthebanana/d0cf5113820cda71b8a7d63a39d100e1#jjtlora-baseline-afbc061).

jacobthebanana committed:
Merge branch 'master' into jjt/lora-baseline

jacobthebanana committed:
Replaced optimizer.load_state_dict with load_sharded_optimizer_state_dict for PEFT optimizer.
Added LoRA/PEFT documentations.

jacobthebanana commented:
Equivalence between 55ec4af and afbc061 ([compare](https://github.com/VectorInstitute/vectorlm/compare/afbc061...55ec4af)) in terms of ability for LoRA to resume from preemption:
- Previous (afbc061): [link](https://api.wandb.ai/links/jacobthebanana/wvsvwar7)
- Updated (55ec4af): [link](https://api.wandb.ai/links/jacobthebanana/o153z2aw)

jacobthebanana committed:
benchmarking: deleted unused TypeVar in parse_benchmark.py

jacobthebanana committed:
Replaced config getattr and hasattr with dict methods.

jacobthebanana committed:
Deleted redundant lora-specific launch scripts.

jacobthebanana committed:
Added launch_benchmark.sh for throughput benchmarks.

jacobthebanana committed:
Benchmark: run `makedirs` only `if __name__ == "__main__"`.

adil-a commented on README.md:
```diff
@@ -61,14 +61,10 @@ We implement several training optimizations that can be reviewed under [`docs/tr
 # Contributors
 
-Adil Asif, Ziwen Han, John Willes, Jacob-Junqi Tian.
+Adil Asif, Ziwen Han, John Willes.
```
Not sure if this was missed in a commit, but feel free to add yourself here.

adil-a commented on docs/reference_throughput.md:
```diff
@@ -0,0 +1,33 @@
+# Reference Throughput
```
Could you make note of the CUDA and PyTorch version here please?

jacobthebanana replied:
Added.

adil-a commented on examples/__init__.py:
Don't need this file.

adil-a commented on profiling/__init__.py:
I think we don't need this file either (since this directory is not really part of the package).

jacobthebanana replied:
We might need to add INP001 to ruff ignore list. Is that okay?

adil-a commented on vectorlm/trainer.py:
```diff
@@ -59,12 +51,14 @@ class Trainer:
 
     """
 
+    peft_method: str | None = None
```
Small nitpick but can we move this into an initialization declaration please? Just to keep style consistent.

jacobthebanana replied:
Sounds good- this actually might make a difference- on the hindsight, future torch versions might create issues when modifying class variables, which are shared across all instances of Trainer (not on other ranks, but might be the case in the future.)

adil-a commented on vectorlm/utils/model_utils.py:
```diff
@@ -27,106 +30,63 @@
-    """
-    Return whether bfloat is supported for the
-    current CUDA device.
+def get_half_precision_model(model: nn.Module) -> nn.Module:
```
Does this need to be a function? It seems to not do anything significant.

jacobthebanana replied:
The initially idea was to switch between bfloat16 and float16 depending on the GPU (it seems that not all GPUs support bfloat16), but at the moment we are using bfloat16 only because:
- LoRA didn't work well with float16.
- Most of our GPUs are compatible with bfloat16.

I will revert this back to shortly- might reopen when we need to support a GPU that isn't compatible with bfloat16.

adil-a commented on vectorlm/utils/model_utils.py:
```diff
@@ -273,19 +272,23 @@ def shard_model(
         low_cpu_mem_usage: Whether to only load model weights on main rank, and
             then scatter them to the other workers.
+        is_lora_enabled: Whether to enable support for LoRA, where only a subset
+            of. parameter tensors requires_grad. Enabling might significantly
```
Incorrect `.` after of.

jacobthebanana replied:
Resolved- let me know if you prefer keeping these lengthy documentations in a separate location.

adil-a reviewed (changes requested):
Just some small final changes requested.

Thanks for all the solid work, Jacob!

I am going to benchmark this branch against master on an A100 node this week. If that final check passes, let's merge. :)

jacobthebanana committed:
Replaced peft class attributes in Trainer with instance attributes.
Added information about benchmarking environment.
Additional formatting fixes.

adil-a reviewed (approved):
LGTM. Throughput seems fine, thanks for the great work Jacob!
